// @neo4j/graphql/dist/schema-model/attribute/Attribute.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Attribute = exports.AttributeType = void 0;
const Error_1 = require("../../classes/Error");
const Annotation_1 = require("../annotation/Annotation");
var AttributeType;
(function (AttributeType) {
    AttributeType["Boolean"] = "Boolean";
    AttributeType["ID"] = "ID";
    AttributeType["String"] = "String";
    AttributeType["Int"] = "Int";
    AttributeType["BigInt"] = "BigInt";
    AttributeType["Float"] = "Float";
    AttributeType["DateTime"] = "DateTime";
    AttributeType["LocalDateTime"] = "LocalDateTime";
    AttributeType["Time"] = "Time";
    AttributeType["LocalTime"] = "LocalTime";
    AttributeType["Date"] = "Date";
    AttributeType["Duration"] = "Duration";
    AttributeType["Point"] = "Point";
    AttributeType["ObjectType"] = "ObjectType";
})(AttributeType || (exports.AttributeType = AttributeType = {}));
class Attribute {
    constructor({ name, annotations, type, isArray, }) {
        this.annotations = {};
        this.name = name;
        this.type = type;
        this.isArray = isArray;
        for (const annotation of annotations) {
            this.addAnnotation(annotation);
        }
    }
    clone() {
        return new Attribute({
            name: this.name,
            annotations: Object.values(this.annotations),
            type: this.type,
            isArray: this.isArray,
        });
    }
    addAnnotation(annotation) {
        const annotationKey = (0, Annotation_1.annotationToKey)(annotation);
        if (this.annotations[annotationKey]) {
            throw new Error_1.Neo4jGraphQLSchemaValidationError(`Annotation ${annotationKey} already exists in ${this.name}`);
        }
        // We cast to any because we aren't narrowing the Annotation type here.
        // There's no reason to narrow either, since we care more about performance.
        this.annotations[annotationKey] = annotation;
    }
}
exports.Attribute = Attribute;
