// @neo4j/cypher-builder/dist/expressions/functions/TemporalFunctions.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.cypherTime = exports.cypherLocalTime = exports.cypherLocalDatetime = exports.cypherDate = exports.cypherDatetime = void 0;
const CypherFunctions_1 = require("./CypherFunctions");
function dateFunction(name, timezone) {
    return new CypherFunctions_1.CypherFunction(name, timezone ? [timezone] : undefined);
}
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/temporal/#functions-datetime)
 * @group Expressions
 * @category Cypher Functions
 */
function cypherDatetime(timezone) {
    return dateFunction("datetime", timezone);
}
exports.cypherDatetime = cypherDatetime;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/temporal/#functions-date)
 * @group Expressions
 * @category Cypher Functions
 * @example
 *
 * Using date without parameters:
 *
 * ```ts
 * Cypher.Date()
 * ```
 *
 * _Cypher:_
 * ```cypher
 * date()
 * ```
 *
 * ---
 *
 * Date with parameters:
 *
 * ```ts
 * Cypher.Date(new Cypher.param('9999-01-01'))
 * ```
 *
 * _Cypher:_
 * ```cypher
 * date($param1)
 * ```
 *
 * _Params:_
 * ```json
 * {
 *   param1: "9999-01-01"
 * }
 * ```
 *
 *
 */
function cypherDate(timezone) {
    return dateFunction("date", timezone);
}
exports.cypherDate = cypherDate;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/temporal/#functions-localdatetime)
 * @group Expressions
 * @category Cypher Functions
 */
function cypherLocalDatetime(timezone) {
    return dateFunction("localdatetime", timezone);
}
exports.cypherLocalDatetime = cypherLocalDatetime;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/temporal/#functions-localdatetime)
 * @group Expressions
 * @category Cypher Functions
 */
function cypherLocalTime(timezone) {
    return dateFunction("localtime", timezone);
}
exports.cypherLocalTime = cypherLocalTime;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/temporal/#functions-time)
 * @group Expressions
 * @category Cypher Functions
 */
function cypherTime(timezone) {
    return dateFunction("time", timezone);
}
exports.cypherTime = cypherTime;
