// @neo4j/graphql/dist/classes/authorization/Neo4jGraphQLAuthorization.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Neo4jGraphQLAuthorization = void 0;
const debug_1 = __importDefault(require("debug"));
const neo4j_driver_1 = require("neo4j-driver");
const constants_1 = require("../../constants");
const jose_1 = require("jose");
const parse_request_token_1 = require("./parse-request-token");
const debug = (0, debug_1.default)(constants_1.DEBUG_AUTH);
class Neo4jGraphQLAuthorization {
    constructor(authorization) {
        this.authorization = authorization;
    }
    async decode(req) {
        const bearerToken = (0, parse_request_token_1.getToken)(req);
        if (!bearerToken) {
            return;
        }
        const token = (0, parse_request_token_1.parseBearerToken)(bearerToken);
        if (!token) {
            throw new neo4j_driver_1.Neo4jError("Unauthenticated", constants_1.AUTH_FORBIDDEN_ERROR);
        }
        try {
            if (this.authorization.verify === false) {
                debug("Skipping verifying JWT as verify is set to false");
                return (0, jose_1.decodeJwt)(token);
            }
            const secret = this.resolveKey(req);
            return await this.verify(token, secret);
        }
        catch (error) {
            debug("%s", error);
            if (error instanceof jose_1.errors.JWSInvalid) {
                throw new neo4j_driver_1.Neo4jError("Unauthenticated", constants_1.AUTH_FORBIDDEN_ERROR);
            }
            throw new neo4j_driver_1.Neo4jError("Forbidden", constants_1.AUTH_FORBIDDEN_ERROR);
        }
    }
    decodeBearerToken(bearerToken) {
        const token = (0, parse_request_token_1.parseBearerToken)(bearerToken);
        if (!token)
            throw new Error();
        return (0, jose_1.decodeJwt)(token);
    }
    resolveKey(req) {
        if (typeof this.authorization.key === "function") {
            return this.authorization.key(req);
        }
        else {
            return this.authorization.key;
        }
    }
    async verify(token, secret) {
        if (typeof secret === "string") {
            debug("Verifying JWT using secret");
            const { payload } = await (0, jose_1.jwtVerify)(token, Buffer.from(secret), this.authorization.verifyOptions);
            return payload;
        }
        debug("Verifying JWKS using url");
        const { url, options } = secret;
        const JWKS = (0, jose_1.createRemoteJWKSet)(new URL(url), options);
        const { payload } = await (0, jose_1.jwtVerify)(token, JWKS, this.authorization.verifyOptions);
        return payload;
    }
}
exports.Neo4jGraphQLAuthorization = Neo4jGraphQLAuthorization;
