// @neo4j/graphql/dist/schema-model/annotation/Annotation.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.annotationToKey = exports.AnnotationsKey = void 0;
const AuthenticationAnnotation_1 = require("./AuthenticationAnnotation");
const AuthorizationAnnotation_1 = require("./AuthorizationAnnotation");
const CypherAnnotation_1 = require("./CypherAnnotation");
const KeyAnnotation_1 = require("./KeyAnnotation");
const SubscriptionsAuthorizationAnnotation_1 = require("./SubscriptionsAuthorizationAnnotation");
var AnnotationsKey;
(function (AnnotationsKey) {
    AnnotationsKey["cypher"] = "cypher";
    AnnotationsKey["authorization"] = "authorization";
    AnnotationsKey["authentication"] = "authentication";
    AnnotationsKey["key"] = "key";
    AnnotationsKey["subscriptionsAuthorization"] = "subscriptionsAuthorization";
})(AnnotationsKey || (exports.AnnotationsKey = AnnotationsKey = {}));
function annotationToKey(ann) {
    if (ann instanceof CypherAnnotation_1.CypherAnnotation)
        return AnnotationsKey.cypher;
    if (ann instanceof AuthorizationAnnotation_1.AuthorizationAnnotation)
        return AnnotationsKey.authorization;
    if (ann instanceof AuthenticationAnnotation_1.AuthenticationAnnotation)
        return AnnotationsKey.authentication;
    if (ann instanceof KeyAnnotation_1.KeyAnnotation)
        return AnnotationsKey.key;
    if (ann instanceof SubscriptionsAuthorizationAnnotation_1.SubscriptionsAuthorizationAnnotation)
        return AnnotationsKey.subscriptionsAuthorization;
    throw new Error("annotation not known");
}
exports.annotationToKey = annotationToKey;
