// @neo4j/graphql/dist/schema/create-relationship-fields/create-relationship-interface-fields.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRelationshipInterfaceFields = void 0;
const graphql_compose_1 = require("graphql-compose");
const constants_1 = require("../../constants");
const directed_argument_1 = require("../directed-argument");
const to_compose_1 = require("../to-compose");
function createRelationshipInterfaceFields({ nodes, rel, composeNode, schemaComposer, sourceName, hasNonGeneratedProperties, anyNonNullRelProperties, }) {
    const refNodes = nodes.filter((x) => rel.interface?.implementations?.includes(x.name));
    const upperFieldName = (0, graphql_compose_1.upperFirst)(rel.fieldName);
    const nestedOperations = new Set(rel.nestedOperations);
    const nodeCreateInput = schemaComposer.getITC(`${sourceName}CreateInput`);
    const nodeUpdateInput = schemaComposer.getITC(`${sourceName}UpdateInput`);
    const connectWhere = schemaComposer.getOrCreateITC(`${rel.typeMeta.name}ConnectWhere`, (tc) => {
        tc.addFields({
            node: `${rel.typeMeta.name}Where!`,
        });
    });
    const connectFieldInput = schemaComposer.getOrCreateITC(`${sourceName}${upperFieldName}ConnectFieldInput`, (tc) => {
        if (schemaComposer.has(`${rel.typeMeta.name}ConnectInput`)) {
            tc.addFields({ connect: `${rel.typeMeta.name}ConnectInput` });
        }
        if (hasNonGeneratedProperties) {
            tc.addFields({ edge: `${rel.properties}CreateInput${anyNonNullRelProperties ? `!` : ""}` });
        }
        tc.addFields({ where: connectWhere });
    });
    const deleteFieldInput = schemaComposer.getOrCreateITC(`${sourceName}${upperFieldName}DeleteFieldInput`, (tc) => {
        if (schemaComposer.has(`${rel.typeMeta.name}DeleteInput`)) {
            tc.addFields({ delete: `${rel.typeMeta.name}DeleteInput` });
        }
        tc.addFields({ where: `${rel.connectionPrefix}${upperFieldName}ConnectionWhere` });
    });
    const disconnectFieldInput = schemaComposer.getOrCreateITC(`${sourceName}${upperFieldName}DisconnectFieldInput`, (tc) => {
        if (schemaComposer.has(`${rel.typeMeta.name}DisconnectInput`)) {
            tc.addFields({ disconnect: `${rel.typeMeta.name}DisconnectInput` });
        }
        tc.addFields({ where: `${rel.connectionPrefix}${upperFieldName}ConnectionWhere` });
    });
    const createFieldInput = schemaComposer.getOrCreateITC(`${sourceName}${upperFieldName}CreateFieldInput`, (tc) => {
        tc.addFields({
            node: `${rel.typeMeta.name}CreateInput!`,
        });
        if (hasNonGeneratedProperties) {
            tc.addFields({
                edge: `${rel.properties}CreateInput${anyNonNullRelProperties ? `!` : ""}`,
            });
        }
    });
    const updateConnectionFieldInput = schemaComposer.getOrCreateITC(`${sourceName}${upperFieldName}UpdateConnectionInput`, (tc) => {
        if (hasNonGeneratedProperties) {
            tc.addFields({ edge: `${rel.properties}UpdateInput` });
        }
        tc.addFields({ node: `${rel.typeMeta.name}UpdateInput` });
    });
    if (nestedOperations.size !== 0 &&
        !(nestedOperations.size === 1 && nestedOperations.has(constants_1.RelationshipNestedOperationsOption.CONNECT_OR_CREATE))) {
        const updateFieldInput = schemaComposer.getOrCreateITC(`${sourceName}${upperFieldName}UpdateFieldInput`);
        if (nestedOperations.has(constants_1.RelationshipNestedOperationsOption.CONNECT)) {
            const nodeConnectInput = schemaComposer.getOrCreateITC(`${sourceName}ConnectInput`);
            nodeConnectInput.addFields({
                [rel.fieldName]: rel.typeMeta.array ? connectFieldInput.NonNull.List : connectFieldInput,
            });
            updateFieldInput.addFields({
                connect: rel.typeMeta.array ? connectFieldInput.NonNull.List : connectFieldInput,
            });
        }
        if (nestedOperations.has(constants_1.RelationshipNestedOperationsOption.DELETE)) {
            const nodeDeleteInput = schemaComposer.getOrCreateITC(`${sourceName}DeleteInput`);
            nodeDeleteInput.addFields({
                [rel.fieldName]: rel.typeMeta.array ? deleteFieldInput.NonNull.List : deleteFieldInput,
            });
            updateFieldInput.addFields({
                delete: rel.typeMeta.array ? deleteFieldInput.NonNull.List : deleteFieldInput,
            });
        }
        if (nestedOperations.has(constants_1.RelationshipNestedOperationsOption.DISCONNECT)) {
            const nodeDisconnectInput = schemaComposer.getOrCreateITC(`${sourceName}DisconnectInput`);
            nodeDisconnectInput.addFields({
                [rel.fieldName]: rel.typeMeta.array ? disconnectFieldInput.NonNull.List : disconnectFieldInput,
            });
            updateFieldInput.addFields({
                disconnect: rel.typeMeta.array ? disconnectFieldInput.NonNull.List : disconnectFieldInput,
            });
        }
        if (nestedOperations.has(constants_1.RelationshipNestedOperationsOption.CREATE)) {
            const nodeRelationInput = schemaComposer.getOrCreateITC(`${sourceName}RelationInput`);
            nodeRelationInput.addFields({
                [rel.fieldName]: rel.typeMeta.array ? createFieldInput.NonNull.List : createFieldInput,
            });
            updateFieldInput.addFields({
                create: rel.typeMeta.array ? createFieldInput.NonNull.List : createFieldInput,
            });
        }
        if (nestedOperations.has(constants_1.RelationshipNestedOperationsOption.UPDATE)) {
            updateFieldInput.addFields({
                update: updateConnectionFieldInput,
            });
        }
        updateFieldInput.addFields({
            where: `${rel.connectionPrefix}${upperFieldName}ConnectionWhere`,
        });
        if (rel.settableOptions.onUpdate) {
            nodeUpdateInput.addFields({
                [rel.fieldName]: rel.typeMeta.array ? updateFieldInput.NonNull.List : updateFieldInput,
            });
        }
    }
    if (!rel.writeonly) {
        const baseNodeFieldArgs = {
            options: `${rel.typeMeta.name}Options`,
            where: `${rel.typeMeta.name}Where`,
        };
        const nodeFieldArgs = (0, directed_argument_1.addDirectedArgument)(baseNodeFieldArgs, rel);
        if (rel.selectableOptions.onRead) {
            composeNode.addFields({
                [rel.fieldName]: {
                    type: rel.typeMeta.pretty,
                    args: nodeFieldArgs,
                    description: rel.description,
                    directives: (0, to_compose_1.graphqlDirectivesToCompose)(rel.otherDirectives),
                },
            });
        }
    }
    if (nestedOperations.has(constants_1.RelationshipNestedOperationsOption.CONNECT) ||
        nestedOperations.has(constants_1.RelationshipNestedOperationsOption.CREATE)) {
        const nodeFieldInput = schemaComposer.getOrCreateITC(`${rel.connectionPrefix}${upperFieldName}FieldInput`, (tc) => {
            if (nestedOperations.has(constants_1.RelationshipNestedOperationsOption.CREATE)) {
                tc.addFields({
                    create: rel.typeMeta.array ? createFieldInput.NonNull.List : createFieldInput,
                });
            }
            if (nestedOperations.has(constants_1.RelationshipNestedOperationsOption.CONNECT)) {
                tc.addFields({
                    connect: rel.typeMeta.array ? connectFieldInput.NonNull.List : connectFieldInput,
                });
            }
        });
        refNodes.forEach((n) => {
            const createName = `${sourceName}${upperFieldName}${n.name}CreateFieldInput`;
            if (!schemaComposer.has(createName)) {
                schemaComposer.getOrCreateITC(createName, (tc) => {
                    tc.addFields({ node: `${n.name}CreateInput!` });
                    if (hasNonGeneratedProperties) {
                        tc.addFields({ edge: `${rel.properties}CreateInput${anyNonNullRelProperties ? `!` : ""}` });
                    }
                });
            }
        });
        // Interface CreateInput does not require relationship input fields
        // These are specified on the concrete nodes.
        if (rel.settableOptions.onCreate && !(composeNode instanceof graphql_compose_1.InterfaceTypeComposer)) {
            nodeCreateInput.addFields({
                [rel.fieldName]: nodeFieldInput,
            });
        }
    }
}
exports.createRelationshipInterfaceFields = createRelationshipInterfaceFields;
