// @neo4j/cypher-builder/dist/expressions/operations/comparison.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.matches = exports.endsWith = exports.startsWith = exports.contains = exports.inOp = exports.isNotNull = exports.isNull = exports.lte = exports.lt = exports.gte = exports.gt = exports.neq = exports.eq = exports.ComparisonOp = void 0;
const CypherASTNode_1 = require("../../CypherASTNode");
/**
 *  @group Internal
 */
class ComparisonOp extends CypherASTNode_1.CypherASTNode {
    constructor(operator, left, right) {
        super();
        this.operator = operator;
        this.leftExpr = left;
        this.rightExpr = right;
    }
    /**
     * @hidden
     */
    getCypher(env) {
        const leftStr = `${this.leftExpr.getCypher(env)} `;
        const rightStr = this.rightExpr ? ` ${this.rightExpr.getCypher(env)}` : "";
        return `${leftStr}${this.operator}${rightStr}`;
    }
}
exports.ComparisonOp = ComparisonOp;
function createOp(op, leftExpr, rightExpr) {
    return new ComparisonOp(op, leftExpr, rightExpr);
}
/** Equality (=) operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function eq(leftExpr, rightExpr) {
    return createOp("=", leftExpr, rightExpr);
}
exports.eq = eq;
/** Inequality (<>) operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function neq(leftExpr, rightExpr) {
    return createOp("<>", leftExpr, rightExpr);
}
exports.neq = neq;
/** Greater Than (>) operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function gt(leftExpr, rightExpr) {
    return createOp(">", leftExpr, rightExpr);
}
exports.gt = gt;
/** Greater Than Equal (>=) operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function gte(leftExpr, rightExpr) {
    return createOp(">=", leftExpr, rightExpr);
}
exports.gte = gte;
/** Less Than (<) operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function lt(leftExpr, rightExpr) {
    return createOp("<", leftExpr, rightExpr);
}
exports.lt = lt;
/** Less Than Equal (<=) operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function lte(leftExpr, rightExpr) {
    return createOp("<=", leftExpr, rightExpr);
}
exports.lte = lte;
/** `IS NULL` operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 * @example
 * ```cypher
 * this0.title IS NULL
 * ```
 */
function isNull(childExpr) {
    return createOp("IS NULL", childExpr);
}
exports.isNull = isNull;
/** `IS NOT NULL` operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 * @example
 * ```cypher
 * this0.title IS NULL
 * ```
 */
function isNotNull(childExpr) {
    return createOp("IS NOT NULL", childExpr);
}
exports.isNotNull = isNotNull;
/** `IN` operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function inOp(leftExpr, rightExpr) {
    return createOp("IN", leftExpr, rightExpr);
}
exports.inOp = inOp;
/** `CONTAINS` operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function contains(leftExpr, rightExpr) {
    return createOp("CONTAINS", leftExpr, rightExpr);
}
exports.contains = contains;
/** `STARTS WITH` operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function startsWith(leftExpr, rightExpr) {
    return createOp("STARTS WITH", leftExpr, rightExpr);
}
exports.startsWith = startsWith;
/** `ENDS WITH` operator
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function endsWith(leftExpr, rightExpr) {
    return createOp("ENDS WITH", leftExpr, rightExpr);
}
exports.endsWith = endsWith;
/** Matching (=~) operator.
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-comparison)
 * @group Operators
 * @category Comparison
 */
function matches(leftExpr, rightExpr) {
    return createOp("=~", leftExpr, rightExpr);
}
exports.matches = matches;
