// @neo4j/graphql/dist/schema-model/relationship/Relationship.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Relationship = void 0;
const classes_1 = require("../../classes");
const upper_first_1 = require("../../utils/upper-first");
class Relationship {
    /**Note: Required for now to infer the types without ResolveTree */
    get connectionFieldTypename() {
        return `${this.source.name}${(0, upper_first_1.upperFirst)(this.name)}Connection`;
    }
    /**Note: Required for now to infer the types without ResolveTree */
    get relationshipFieldTypename() {
        return `${this.source.name}${(0, upper_first_1.upperFirst)(this.name)}Relationship`;
    }
    constructor({ name, type, attributes = [], source, target, direction, }) {
        this.attributes = new Map();
        this.type = type;
        this.source = source;
        this.target = target;
        this.name = name;
        this.direction = direction;
        for (const attribute of attributes) {
            this.addAttribute(attribute);
        }
    }
    clone() {
        return new Relationship({
            name: this.name,
            type: this.type,
            attributes: Array.from(this.attributes.values()).map((a) => a.clone()),
            source: this.source,
            target: this.target,
            direction: this.direction,
        });
    }
    addAttribute(attribute) {
        if (this.attributes.has(attribute.name)) {
            throw new classes_1.Neo4jGraphQLSchemaValidationError(`Attribute ${attribute.name} already exists in ${this.name}.`);
        }
        this.attributes.set(attribute.name, attribute);
    }
    findAttribute(name) {
        return this.attributes.get(name);
    }
}
exports.Relationship = Relationship;
