// @neo4j/graphql/dist/classes/SubscriptionDirective.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionDirective = void 0;
const subscription_1 = require("../graphql/directives/subscription");
class SubscriptionDirective {
    constructor(operations) {
        const operationsSet = new Set(operations);
        this.create = operationsSet.has(subscription_1.SubscriptionOperations.CREATE);
        this.update = operationsSet.has(subscription_1.SubscriptionOperations.UPDATE);
        this.delete = operationsSet.has(subscription_1.SubscriptionOperations.DELETE);
        this.createRelationship = operationsSet.has(subscription_1.SubscriptionOperations.CREATE_RELATIONSHIP);
        this.deleteRelationship = operationsSet.has(subscription_1.SubscriptionOperations.DELETE_RELATIONSHIP);
    }
}
exports.SubscriptionDirective = SubscriptionDirective;
